### PR TITLE
add fill_diagonal function

### LINKED ISSUE #21796: Add torch.fill_diagonal
## 🚀 Feature
Implement torch.fill_diagonal based on numpy.fill_diagonal

See https://docs.scipy.org/doc/numpy/reference/generated/numpy.fill_diagonal.html

## Motivation
Alex asked me how to do this today.

## Alternatives
The following works but isn't obvious and may not work for non-strided Tensors:

```python
torch.diagonal(x).fill_(value)
```

### PR DESCRIPTION
**Pull request by shihongzhi**
Fixes #21796

### CODE REVIEW

**Review by colesbury — CHANGES_REQUESTED** [review]
Thanks for taking on this issue! This needs a few changes before it can be merged:

1) The API should match the NumPy API: https://docs.scipy.org/doc/numpy/reference/generated/numpy.fill_diagonal.html
2) The function needs API documentation (in [torch/_torch_docs.py](https://github.com/pytorch/pytorch/blob/master/torch/_torch_docs.py))
3) The function needs unit tests. (In test/test_torch.py)

Once the PR is ready for review, please click "dismiss review" on this

File: aten/src/ATen/native/TensorFactories.cpp [hunk 1]
```diff
@@ -249,6 +249,18 @@ Tensor full_like(const Tensor& self, Scalar fill_value, const TensorOptions& opt
   return native::full(self.sizes(), fill_value, options);
 }
 
+// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ fill diagonal ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Tensor fill_diagonal(const Tensor& self, Scalar value, int64_t offset, int64_t dim1_, int64_t dim2_) {
+  int64_t nDims = self.dim();
```
Comment (SsnL) [inline aten/src/ATen/native/TensorFactories.cpp]: doesn't tensor.diagonal already do these checks?

Comment (shihongzhi) [reply aten/src/ATen/native/TensorFactories.cpp]: yes. I remove these checks, and change api to match Numpy api. Thank you

**Comment by apaszke** [comment]
To add to what @colesbury said, the name will also need to be changed to have a trailing underscore, since it modifies the input tensor.

**Comment by shihongzhi** [comment]
@apaszke  Already add trailing underscore to function name. Thank you

@colesbury  I had add docs&unit tests and match this API to Numpy API

File: aten/src/ATen/native/native_functions.yaml [hunk 2]
```diff
@@ -614,6 +614,9 @@
 - func: diagonal(Tensor(a) self, int offset=0, int dim1=0, int dim2=1) -> Tensor(a)
   variants: function, method
 
+- func: fill_diagonal_(Tensor(a!) self, Scalar fill_value, bool wrap=False) -> Tensor(a!)
+  variants: function, method
```
Comment (colesbury) [inline aten/src/ATen/native/native_functions.yaml]: Do we want this to be a function on torch? Originally, we only had in-place operations as methods on torch.Tensor but we've seemed to relax that (perhaps unintentionally)

cc @gchanan @apaszke

File: test/test_torch.py [hunk 3]
```diff
@@ -12179,6 +12179,23 @@ def test_python_types(self):
         c2 = torch.tensor([True, False], dtype=bool)
         self.assertEqual(c1.dtype, c2.dtype)
 
+    def test_fill_diagonal(self):
```
Comment (colesbury) [inline test/test_torch.py]: Please add test cases for:

1) A 3-d or larger tensor
2) A non-contiguous tensor

File: torch/_torch_docs.py [hunk 4]
```diff
@@ -5815,6 +5815,47 @@ def merge_dicts(*dicts):
...
+           r"""
+fill_diagonal_(input, fill_value, wrap=False) -> Tensor
+
+Fill the main diagonal of the given tensor of any dimensionality.
+This function modifies the input tensor in-place, and return the input tensor.
+
+Arguments:
+    input (Tensor): at least 2-dims. When dims>2, all dimensions of input must be of equal length
+    fill_value (Scalar): the fill value
+    wrap: the diagonal 'wrapped' after N columns for tall matrices.
```
Comment (colesbury) [inline torch/_torch_docs.py]: wrap (bool):

(add type)

File: torch/_torch_docs.py [hunk 5]
```diff
@@ -5815,6 +5815,47 @@ def merge_dicts(*dicts):
     {requires_grad}
 """.format(**factory_like_common_args))
 
+
+add_docstr(torch.fill_diagonal_,
+           r"""
+fill_diagonal_(input, fill_value, wrap=False) -> Tensor
+
+Fill the main diagonal of the given tensor of any dimensionality.
+This function modifies the input tensor in-place, and return the input tensor.
```
Comment (colesbury) [inline torch/_torch_docs.py]: return -> returns

File: aten/src/ATen/native/TensorFactories.cpp [hunk 6]
```diff
@@ -249,6 +249,51 @@ Tensor full_like(const Tensor& self, Scalar fill_value, const TensorOptions& opt
   return native::full(self.sizes(), fill_value, options);
 }
 
+// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ fill diagonal ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Tensor& fill_diagonal_(Tensor& self, Scalar fill_value, bool wrap) {
+  int64_t nDims = self.dim();
+  TORCH_CHECK(nDims >= 2, "dimensions must large then 1 ");
```
Comment (colesbury) [inline aten/src/ATen/native/TensorFactories.cpp]: large -> larger

remove trailing space after `1`

**Review by colesbury — COMMENTED** [review]
Nice! Looking pretty good

File: aten/src/ATen/native/native_functions.yaml (hunk 2 above)
Comment (gchanan) [reply aten/src/ATen/native/native_functions.yaml]: I like in-place ops being method only.

Comment (shihongzhi) [reply aten/src/ATen/native/native_functions.yaml]: I remove 'function'

Comment (apaszke) [reply aten/src/ATen/native/native_functions.yaml]: Yes, it should only be a method.

File: torch/_tensor_docs.py [hunk 7]
```diff
@@ -716,6 +716,46 @@ def add_docstr_all(method, docstr):
 See :func:`torch.diagonal`
 """)
 
+add_docstr_all('fill_diagonal_',
+               r"""
+fill_diagonal_(fill_value, wrap=False) -> Tensor
+
+Fill the main diagonal of the given tensor of any dimensionality.
```
Comment (gchanan) [inline torch/_tensor_docs.py]: I'm having a bit of trouble parsing the description -- do you mean "Fill the main diagonal of a tensor that has at least 2-dimensions?"

Comment (shihongzhi) [reply torch/_tensor_docs.py]: sorry, the description is so confused. I'm changed this.

**Comment by shihongzhi** [comment]
@colesbury I fixed those bug. It's ready for review again.

File: docs/source/torch.rst [hunk 8]
```diff
@@ -50,6 +50,7 @@ Creation Ops
 .. autofunction:: empty_like
 .. autofunction:: full
 .. autofunction:: full_like
+.. autofunction:: fill_diagonal_
```
Comment (colesbury) [inline docs/source/torch.rst]: I think this needs to be removed since it's no longer on the torch package

File: test/test_torch.py [hunk 9]
```diff
@@ -12179,6 +12179,52 @@ def test_python_types(self):
...
+
+        e1 = torch.rand(7, 3, 3)[:, 1, ...]
+        e2 = e1.clone()
+        for i in range(3):
+            e2[i][i] = v
+            e2[i + 4][i] = v
+        e1.fill_diagonal_(v, wrap=True)
+        self.assertEqual(e1, e2)
+
+
```
Comment (colesbury) [inline test/test_torch.py]: please remove the extra whitespace lines

**Review by colesbury — CHANGES_REQUESTED** [review]
Some of the continuous builds are failing (`test_torch (__main__.TestDocCoverage)`) I think because of the function listed in torch.rst but it's not in the torch package.

Otherwise looks good to go

**Comment by shihongzhi** [comment]
@colesbury Thank you.  I removed `fill_diagonal_` in torch.rst

**Comment by shihongzhi** [comment]
@colesbury I think this's ready for merge

**Comment by colesbury** [comment]
Thanks @shihongzhi!

**Review by colesbury — APPROVED** [review]

**Comment by xuhdev** [comment]
Can we move the definition to a different file? It seems to me that fill_diagonal is not a factory. Perhaps create Fill.cpp and move both `fill_diagonal` and `fill` to it? I don't think this is urgent for this PR -- we can do this after this PR is merged.

**Comment by colesbury** [comment]
@xuhdev yes, that seems reasonable. Do you want to make the change after this lands?

**Comment by xuhdev** [comment]
@colesbury See #22758

**Pull request closed without merging**